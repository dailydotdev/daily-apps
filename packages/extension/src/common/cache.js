export const STATE_KEY = 'state';
export const ANALYTICS_ID_KEY = 'analyticsId';
export const ANALYTICS_CONSENT_KEY = 'analyticsConsent';
export const TERMS_CONSENT_KEY = 'termsConsent';
export const DATA_VERSION_KEY = 'dataVersion';
export const CURRENT_DATA_VERSION = 2;

export function setCache(key, value) {
  return browser.storage.local.set({ [key]: value });
}

export async function getCache(key, defaultValue) {
  const res = await browser.storage.local.get(key);
  if (!(Number.isNaN(res[key]) || res[key] === undefined)) {
    return res[key];
  }

  return defaultValue;
}

export function removeCache(key) {
  return browser.storage.local.remove(key);
}
