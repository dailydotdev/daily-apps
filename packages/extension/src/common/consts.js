export const CRITICAL_FETCH_STAGE = 1;
export const OPERATIONAL_FETCH_STAGE = 2;
export const ENGAGEMENT_FETCH_STAGE = 4;
