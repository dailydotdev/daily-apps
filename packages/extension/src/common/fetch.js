// eslint-disable-next-line import/prefer-default-export
export const fetchTimeout = (url, ms, { signal, ...options } = {}) => {
  const controller = new AbortController();
  const promise = fetch(url, { signal: controller.signal, ...options });
  if (signal) {
    signal.addEventListener('abort', () => controller.abort());
  }
  const timeout = setTimeout(() => controller.abort(), ms);
  return promise.finally(() => clearTimeout(timeout));
};
