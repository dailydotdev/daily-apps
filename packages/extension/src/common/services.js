import {
  AuthServiceImpl, ContentServiceImpl, MonetizationServiceImpl, ProfileServiceImpl,
} from '@daily/services';

const app = 'extension';
// eslint-disable-next-line import/prefer-default-export
export const contentService = new ContentServiceImpl(process.env.VUE_APP_API_URL, 30, app);
export const monetizationService = new MonetizationServiceImpl(process.env.VUE_APP_API_URL, app);
export const profileService = new ProfileServiceImpl(process.env.VUE_APP_API_URL, app);
export const authService = new AuthServiceImpl(process.env.VUE_APP_API_URL, app);
