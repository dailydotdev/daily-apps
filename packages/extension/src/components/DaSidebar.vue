<template>
  <aside class="sidebar" :class="{opened}"
         @mouseleave="close">
    <div class="sidebar__wrapper">
      <div class="sidebar__filter">
        <da-mode-switch class="sidebar__filter_switch"
                        first-icon="link" second-icon="hashtag"
                        :checked="filterChecked" @toggle="toggleFilter($event)"/>
      </div>

      <div class="sidebar__element sidebar__search btn btn-menu"
           :class="{selected: searchFocused}">
        <svgicon name="magnifying" class="sidebar__element__image"/>
        <input class="sidebar__input" type="text" :placeholder="searchPlaceholder" ref="search"
               v-on="{ input: filterChecked ? searchTags : searchPublications }"
               @focus="searchFocused = true" @blur="searchFocused = false">
      </div>

      <div class="sidebar__tags" v-if="filterChecked">
        <div class="sidebar__enabled">
          <div class="sidebar__header">My tags</div>
          <div class="sidebar__element btn btn-menu"
               v-for="item in enabledTags" :key="item.name">
            <button class="sidebar__element__button"
                    title="Click to view only this tag" @click.prevent="viewTag(item)">
              <span class="invert sidebar__tag shadow1 text-overflow">#{{item.name}}</span>
            </button>
            <button class="sidebar__element__button-hidden btn-icon"
                    title="Remove this tag" @click.prevent.stop="setEnableTag(item, false)">
              <svgicon name="x"/>
            </button>
          </div>
        </div>
        <div class="sidebar__disabled" v-if="disabledTags.length > 0">
          <div class="sidebar__header">{{query.length ? 'More tags' : 'Hot tags'}}</div>
          <div class="sidebar__element off btn btn-menu"
               v-for="item in disabledTags" :key="item.name">
            <button class="sidebar__element__button"
                    title="Click to view only this tag" @click.prevent="viewTag(item)">
              <span class="text-overflow">#{{item.name}}</span>
            </button>
            <button class="sidebar__element__button-hidden show btn-icon"
                    title="Add this tag"
                    @click.prevent.stop="setEnableTag(item, true)">
              <svgicon name="plus"/>
            </button>
          </div>
        </div>
      </div>

      <div class="sidebar__sources" v-else>
        <div class="sidebar__enabled">
          <div class="sidebar__header">My sources</div>
          <button class="sidebar__element sidebar__sources__act-req btn btn-menu"
                  @click.prevent="activateRequest" v-if="!requestActive">
            <svgicon name="plus" class="sidebar__element__image"/>
            <span>Request source</span>
          </button>
          <form class="sidebar__element btn btn-menu selected" v-else
                @click.prevent="$refs.request.focus()" ref="form">
            <button type="button" class="btn-icon btn-small" @click="cancelRequest">
              <svgicon name="x" class="sidebar__element__image"/>
            </button>
            <input class="sidebar__input" type="url" placeholder="Paste URL" required
                   ref="request" @input="updateFormValidity">
            <button type="submit" class="sidebar__sources__submit btn btn-invert"
                    :disabled="disableSubmit" @click.prevent="submitRequest">
              <svgicon name="v" class="invert"/>
            </button>
            <transition name="bouncy-flip">
              <div class="sidebar__sources__error nuggets" v-if="submitError">
                Something went wrong, try again later
              </div>
            </transition>
          </form>
          <div class="sidebar__element btn btn-menu" v-for="item in enabledPubs"
               :key="item.id">
            <button class="sidebar__element__button"
                    title="Click to view only this source" @click.prevent="viewPublication(item)">
              <img :data-src="item.image" :alt="item.name"
                   class="sidebar__element__image lazyload"/>
              <span class="text-overflow">{{item.name}}</span>
            </button>
            <button class="sidebar__element__button-hidden btn-icon"
                    title="Remove this source"
                    @click.prevent.stop="setEnablePublication(item, false)">
              <svgicon name="x"/>
            </button>
          </div>
        </div>
        <div class="sidebar__disabled" v-if="disabledPubs.length > 0">
          <div class="sidebar__header">More sources</div>
          <div class="sidebar__element off btn btn-menu"
               v-for="item in disabledPubs" :key="item.id">
            <button class="sidebar__element__button"
                    title="Click to view only this source"
                    @click.prevent="viewPublication(item)">
              <img :data-src="item.image" :alt="item.name"
                   class="sidebar__element__image lazyload"/>
              <span class="text-overflow">{{item.name}}</span>
            </button>
            <button class="sidebar__element__button-hidden show btn-icon"
                    title="Add this source"
                    @click.prevent.stop="setEnablePublication(item, true)">
              <svgicon name="plus"/>
            </button>
          </div>
        </div>
      </div>

    </div>
  </aside>
</template>

<script>
import 'lazysizes';
import { mapState, mapActions, mapGetters } from 'vuex';
import DaModeSwitch from '@daily/components/src/components/DaModeSwitch.vue';
import { contentService } from '../common/services';

export default {
  name: 'DaSidebar',

  components: {
    DaModeSwitch,
  },

  props: {
    disabled: {
      type: Boolean,
      default: false,
    },
  },

  data() {
    return {
      opened: false,
      filterChecked: false,
      requestActive: false,
      disableSubmit: true,
      submitError: false,
      searchedTags: [],
      searchFocused: false,
      query: '',
    };
  },

  computed: {
    ...mapState('feed', ['publications', 'tags']),
    ...mapGetters('user', ['isLoggedIn']),
    enabledPubs() {
      return this.filteredPublications.filter(x => x.enabled);
    },
    disabledPubs() {
      return this.filteredPublications.filter(x => !x.enabled);
    },
    enabledTags() {
      return this.activeTags.filter(x => x.enabled);
    },
    disabledTags() {
      return this.activeTags.filter(x => !x.enabled);
    },
    activeTags() {
      if (this.query.length) {
        return this.searchedTags.map((t) => {
          const found = this.tags.find(t2 => t2.name === t.name);
          const enabled = !!(found && found.enabled);
          return { name: t.name, enabled };
        });
      }
      return this.tags;
    },
    searchPlaceholder() {
      return `Search ${this.filterChecked ? 'Tags' : 'Sources'}`;
    },
    filteredPublications() {
      if (!this.query.length) {
        return this.publications;
      }

      return this.publications.filter(
        pub => pub.name.toLowerCase().includes(this.query.toLowerCase()),
      );
    },
  },

  methods: {
    open() {
      if (this.opened || this.disabled) {
        return;
      }

      ga('send', 'event', 'Sidebar', 'Toggle', 'Open');
      this.setOpened(true);
    },
    close() {
      if (!this.opened) {
        return;
      }

      ga('send', 'event', 'Sidebar', 'Toggle', 'Close');
      this.resetRequest();
      this.setOpened(false);
    },
    toggleFilter(checked) {
      ga('send', 'event', 'Sidebar', 'Filter', checked ? 'Tags' : 'Publications');
      this.filterChecked = checked;
      this.query = '';

      this.$refs.search.value = '';
    },
    setEnablePublication(pub, enabled) {
      ga('send', 'event', 'Publications', 'Toggle', enabled ? 'Check' : 'Uncheck');
      const index = this.publications.findIndex(p => p.id === pub.id);
      this.$store.dispatch('feed/setEnablePublication', { index, enabled })
      // TODO: handle error
      // eslint-disable-next-line
                .catch(console.error);
    },
    // eslint-disable-next-line no-unused-vars
    viewPublication(pub) {
      ga('send', 'event', 'Publications', 'Single');
      this.setFilter({ type: 'publication', info: pub });
    },
    activateRequest() {
      if (this.isLoggedIn) {
        ga('send', 'event', 'Request Source', 'Activate');
        this.requestActive = true;
        this.$nextTick(() => {
          this.$refs.request.focus();
        });
      } else {
        this.$emit('login');
      }
    },
    cancelRequest() {
      ga('send', 'event', 'Request Source', 'Cancel');
      this.resetRequest();
    },
    resetRequest() {
      if (this.requestActive) {
        this.requestActive = false;
        this.$refs.request.value = '';
        this.submitError = false;
      }
    },
    async submitRequest() {
      ga('send', 'event', 'Request Source', 'Submit');
      try {
        this.submitError = false;
        await contentService.requestPublication(this.$refs.request.value);
        this.resetRequest();
        this.$emit('requested-source');
      } catch {
        this.submitError = true;
      }
    },
    setEnableTag(tag, enabled) {
      ga('send', 'event', 'Tags', 'Toggle', enabled ? 'Check' : 'Uncheck');
      this.$store.dispatch('feed/setEnableTag', { tag, enabled })
      // TODO: handle error
      // eslint-disable-next-line
                .catch(console.error);
    },
    viewTag(tag) {
      ga('send', 'event', 'Tags', 'Single');
      this.setFilter({ type: 'tag', info: tag });
    },
    updateFormValidity() {
      this.disableSubmit = !this.$refs.form.checkValidity();
      this.submitError = false;
    },
    async searchTags(event) {
      if (!this.query.length) {
        ga('send', 'event', 'Tags', 'Search');
      }
      const query = event.target.value;
      this.query = query;
      if (!query.length) {
        this.searchedTags = [];
      } else {
        try {
          const res = await contentService.searchTags(query);
          this.searchedTags = res.hits;
        } catch (err) {
          // TODO: handle error
          // eslint-disable-next-line
                    console.error(err);
        }
      }
    },
    setOpened(opened) {
      setTimeout(() => {
        this.opened = opened;
      });
    },

    ...mapActions({
      setFilter: 'feed/setFilter',
    }),

    searchPublications(ev) {
      this.query = ev.target.value.trim();
    },
  },

  mounted() {
        import('@daily/components/icons/link');
        import('@daily/components/icons/hashtag');
        import('@daily/components/icons/plus');
        import('@daily/components/icons/x');
        import('@daily/components/icons/v');
        import('@daily/components/icons/magnifying');
  },
};
</script>

<style>
.sidebar {
  display: block;
  position: fixed;
  left: 0;
  top: 0;
  width: 264px;
  height: 100%;
  overflow: hidden;
  background: var(--theme-background-primary);
  border-right: 1px solid var(--theme-background-primary);
  box-shadow: var(--theme-shadow-offset) 0 16px 0 rgba(0, 0, 0, .1);
  transform: translateX(-100%);
  transition: transform 0.15s linear;
  will-change: transform;
  z-index: 20;
  contain: layout paint size;

  &.opened {
    transform: none;
  }

  & .text-overflow {
    display: block;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    min-width: 0;
  }
}

.sidebar__wrapper {
  width: 100%;
  height: 100%;
  overflow-y: scroll;

  &:after {
    content: '';
    position: absolute;
    right: 0;
    top: 0;
    width: 8px;
    height: 100%;
    background: var(--theme-background-highlight);
    z-index: -1;
  }

  &::-webkit-scrollbar {
    width: 8px;
    background-color: transparent;
  }

  &::-webkit-scrollbar-track {
    background: transparent;
  }

  &::-webkit-scrollbar-thumb {
    border-radius: 4px;
    background: var(--theme-disabled);
  }

  &::-webkit-scrollbar-thumb:hover {
    background: var(--theme-secondary);
  }

  &::-webkit-scrollbar-thumb:active {
    background: var(--theme-primary);
  }
}

.sidebar__filter {
  display: flex;
  padding-top: 48px;
  background: var(--theme-background-highlight);
}

.sidebar__filter_switch {
  margin: 26px auto;
}

.sidebar__sources .sidebar__element,
.sidebar__element.sidebar__search {
  height: 44px;
}

.sidebar__tags .sidebar__element {
  height: 40px;
}

.sidebar__element {
  --button-color: var(--theme-secondary);
  text-transform: none;
}

.sidebar__element, .sidebar__element .sidebar__element__button {
  & {
    @mixin micro1;
  }
}

.sidebar__element .sidebar__element__button {
  display: flex;
  flex-direction: row;
  align-items: center;
  border: none;
  background: none;
  color: var(--button-color);
  cursor: pointer;

  & {
    @mixin micro1;
  }
}

.sidebar__element {
  width: 100%;
  padding: 0 16px;

  & .sidebar__element__button {
    display: flex;
    height: 100%;
    padding: 0;
    min-width: 0;
    flex: 1;
  }

  &.off:hover {
    & .sidebar__element__button {
      color: var(--theme-primary);
    }
  }

  &:hover {
    & .sidebar__element__button-hidden {
      display: flex;
    }
  }

  & .sidebar__element__image {
    width: 24px;
    height: 24px;
    border-radius: 4px;
  }

  & > .sidebar__element__image:first-child,
  & .sidebar__element__button .sidebar__element__image {
    margin-left: 0;
    margin-right: 16px;
  }

  & .sidebar__element__button-hidden {
    display: none;
    margin-left: auto;

    & .svg-icon {
      width: 20px;
      height: 20px;
    }

    &.show {
      display: flex;
    }
  }
}

form.sidebar__element {
  position: relative;
  cursor: text;
  overflow: visible;
  z-index: 2;

  & .sidebar__element__image {
    margin-right: 0;
  }

  & .sidebar__sources__submit {
    width: 30px;
    height: 30px;
    padding: 3px;
    border-radius: 8px;

    &:before {
      transition: background-color 0.1s linear;
    }

    & .svg-icon:first-child {
      margin: 0;
    }
  }
}

.sidebar__sources__error {
  position: absolute;
  display: flex;
  left: 0;
  right: 0;
  top: 100%;
  width: 198px;
  height: 48px;
  align-items: center;
  justify-content: center;
  margin: -4px auto 0;
  padding: 0 12px;
  border-radius: 4px;
  color: var(--theme-primary);
  background: var(--color-ketchup-80);
  box-shadow: 0 var(--theme-shadow-offset) 16px 4px rgba(0, 0, 0, 0.32);
  will-change: transform;
}

.sidebar__header {
  height: auto;
  margin: 0 16px 8px;
  padding: 0;
  color: var(--theme-disabled);

  & {
    @mixin lil2;
  }
}

.sidebar__enabled {
  border-radius: 0 0 8px 8px;
  padding-bottom: 8px;
  background: var(--theme-background-highlight);
}

.sidebar__disabled {
  padding: 16px 0;
}

.sidebar__tags .sidebar__disabled,
.sidebar__sources .sidebar__disabled {
  padding-bottom: 44px;
}

.sidebar__input {
  width: 100%;
  flex: 1;
  margin: 0 8px;
  color: var(--theme-primary);
  background: none;
  border: none;
  caret-color: var(--color-water-60);
  min-width: 0;

  & {
    @mixin micro1;
  }

  &:focus {
    outline: 0;
  }

  &::placeholder {
    color: var(--theme-secondary);
  }
}

.sidebar__tag {
  padding: 2px 8px;
  color: var(--theme-primary);
  background: var(--theme-background-highlight);
  border-radius: 4px;
}

.sidebar__element.sidebar__search {
  position: fixed;
  bottom: 0;
  background: var(--theme-background-secondary);
  border-top: 1px solid var(--theme-separator);
  z-index: 2;
  --button-border-radius: 0;

  & .sidebar__element__image {
    margin: 0;
  }

  & .sidebar__input {
    margin-right: 0;
  }
}

.bouncy-flip-enter-active,
.bouncy-flip-leave-active {
  transform-origin: 50% 0;
}

.bouncy-flip-enter-active {
  animation: flipInX 0.5s;
}

.bouncy-flip-leave-active {
  animation: flipInXSimple 0.3s reverse;
}

@keyframes flipInX {
  0% {
    transform: perspective(400px) rotate3d(1, 0, 0, -90deg);
    transition-timing-function: ease-in;
  }

  40% {
    transform: perspective(400px) rotate3d(1, 0, 0, 20deg);
    transition-timing-function: ease-out;
  }

  60% {
    transform: perspective(400px) rotate3d(1, 0, 0, -10deg);
    transition-timing-function: ease-in;
    opacity: 1;
  }

  80% {
    transform: perspective(400px) rotate3d(1, 0, 0, 5deg);
    transition-timing-function: ease-out;
  }

  100% {
    transform: perspective(400px);
  }
}

@keyframes flipInXSimple {
  0% {
    transform: perspective(400px) rotate3d(1, 0, 0, -90deg);
    transition-timing-function: ease-in;
  }
  100% {
    transform: perspective(400px);
  }
}
</style>
