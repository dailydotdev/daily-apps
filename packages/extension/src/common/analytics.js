import { version } from './config';
import { browserName } from './browser';
import { getCache, ANALYTICS_ID_KEY } from './cache';

const queue = [];

const initialize = async () => {
  const clientId = await getCache(ANALYTICS_ID_KEY);
  if (window.location.protocol.indexOf('moz-extension') > -1) {
    let page;

    const getMessage = (action, type, ...args) => {
      if (action !== 'send') return null;

      const prefix = `v=1&tid=${process.env.VUE_APP_GA}&cid=${clientId}&aip=1&dp=${page}`;
      if (type === 'event') {
        return `${prefix}&t=event&ec=${encodeURIComponent(args[0])}&ea=${encodeURIComponent(args[1])}&el=${encodeURIComponent(args[2])}`;
      }
      if (type === 'pageview') {
        return `${prefix}&t=pageview`;
      }
      if (type === 'timing') {
        return `${prefix}&t=timing&utc=${encodeURIComponent(args[0])}&utv=${encodeURIComponent(args[1])}&utt=${encodeURIComponent(args[2])}`;
      }
      if (type === 'exception') {
        return `${prefix}&t=exception&exd=${encodeURIComponent(args[0].exDescription)}&exf=${args[0].exFatal ? 1 : 0}`;
      }

      return null;
    };

    window.ga = (action, type, ...args) => {
      if (action === 'set') {
        if (type === 'page') {
          page = encodeURIComponent(args[0]);
        }
        return;
      }

      const request = new XMLHttpRequest();
      const message = getMessage(action, type, ...args);
      if (message) {
        request.withCredentials = true;
        request.open('POST', 'https://www.google-analytics.com/collect', true);
        request.send(message);
      }
    };

    queue.forEach(args => window.ga(...args));
  } else {
    /* eslint-disable */
    (function (i, s, o, g, r, a, m) {
      a = s.createElement(o),
        m = s.getElementsByTagName(o)[0];
      a.async = 1;
      a.src = g;
      m.parentNode.insertBefore(a, m);
    })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');
    /* eslint-enable */
    ga('create', process.env.VUE_APP_GA, { clientId });
    ga('set', 'checkProtocolTask', () => {
    });
    ga('require', 'displayfeatures');
  }
};

if (window.location.protocol.indexOf('moz-extension') > -1) {
  window.ga = (...args) => queue.push(args);
} else {
  /* eslint-disable */
  (function (i, s, o, g, r) {
    i['GoogleAnalyticsObject'] = r;
    i[r] = i[r] || function () {
      (i[r].q = i[r].q || []).push(arguments);
    }, i[r].l = 1 * new Date();
  })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');
  /* eslint-enable */
}

export default function (consent) {
  if (consent) {
    return initialize();
  }

  window.ga = () => {
  };
  return Promise.resolve();
}

export const trackPageView = (page) => {
  const prefix = browserName ? '/extension' : '/';
  const suffix = browserName ? `v=${version}&b=${browserName}` : `v=${version}`;
  ga('set', 'page', `${prefix}${page}?${suffix}`);
  ga('send', 'pageview');
};
