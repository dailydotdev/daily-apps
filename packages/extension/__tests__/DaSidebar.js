import { shallowMount, mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import svgicon from 'vue-svgicon';
import VTooltip from 'v-tooltip';
import DaModeSwitch from '@daily/components/src/components/DaModeSwitch.vue';
import DaSidebar from '../src/components/DaSidebar.vue';
import { contentService } from '../src/common/services';

jest.mock('../src/common/services', () => ({
  contentService: {
    searchTags: jest.fn(),
    requestPublication: jest.fn(),
  },
}));

const REQUEST_URL = 'https://dailynow.co';

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(svgicon);
localVue.use(VTooltip);
localVue.component('da-mode-switch', DaModeSwitch);

let feed;
let user;
let store;

beforeEach(() => {
  window.ga = () => {
  };

  feed = {
    namespaced: true,
    state: {
      publications: [{
        'id': 'airbnb',
        'name': 'Airbnb Engineering',
        'image': 'https://res.cloudinary.com/daily-now/image/upload/t_logo,f_auto/v1/logos/airbnb',
        'enabled': true,
        'twitter': 'AirbnbEng'
      }, {
        'id': 'alligator',
        'name': 'Alligator',
        'image': 'https://res.cloudinary.com/daily-now/image/upload/t_logo,f_auto/v1/logos/alligator',
        'enabled': false,
        'twitter': 'alligatorio'
      }, {
        'id': 'angular',
        'name': 'Angular',
        'image': 'https://res.cloudinary.com/daily-now/image/upload/t_logo,f_auto/v1/logos/angular',
        'enabled': false,
        'twitter': 'angular'
      }, {
        'id': 'aws',
        'name': 'AWS',
        'image': 'https://res.cloudinary.com/daily-now/image/upload/t_logo,f_auto/v1/logos/aws',
        'enabled': true,
        'twitter': 'awscloud'
      }],
      tags: [{
        'name': 'javascript',
        'enabled': true,
      }, {
        'name': 'linux',
        'enabled': true,
      }, {
        'name': 'startup',
        'enabled': false,
      }, {
        'name': 'product',
        'enabled': false,
      }],
    },
    mutations: {
      setPublications: jest.fn(),
      setTags: jest.fn(),
    },
    actions: {
      setFilter: jest.fn(),
      setEnablePublication: jest.fn(),
      setEnableTag: jest.fn(),
    },
  };

  user = {
    namespaced: true,
    state: {
      profile: null,
    },
    getters: {
      isLoggedIn: state => !!state.profile,
    },
  };

  store = new Vuex.Store({
    modules: { feed, user },
  });
});

it('should set enabledPubs according to state', () => {
  const wrapper = shallowMount(DaSidebar, { store, localVue });
  expect(wrapper.vm.enabledPubs).toEqual([{
    'id': 'airbnb',
    'name': 'Airbnb Engineering',
    'image': 'https://res.cloudinary.com/daily-now/image/upload/t_logo,f_auto/v1/logos/airbnb',
    'enabled': true,
    'twitter': 'AirbnbEng'
  }, {
    'id': 'aws',
    'name': 'AWS',
    'image': 'https://res.cloudinary.com/daily-now/image/upload/t_logo,f_auto/v1/logos/aws',
    'enabled': true,
    'twitter': 'awscloud'
  }]);
});

it('should set disabledPubs according to state', () => {
  const wrapper = shallowMount(DaSidebar, { store, localVue });
  expect(wrapper.vm.disabledPubs).toEqual([{
    'id': 'alligator',
    'name': 'Alligator',
    'image': 'https://res.cloudinary.com/daily-now/image/upload/t_logo,f_auto/v1/logos/alligator',
    'enabled': false,
    'twitter': 'alligatorio'
  }, {
    'id': 'angular',
    'name': 'Angular',
    'image': 'https://res.cloudinary.com/daily-now/image/upload/t_logo,f_auto/v1/logos/angular',
    'enabled': false,
    'twitter': 'angular'
  }]);
});

it('should commit "setEnablePublication" when removing publication', (done) => {
  const wrapper = mount(DaSidebar, { store, localVue });
  wrapper.vm.filterChecked = false;
  wrapper.vm.$nextTick(() => {
    wrapper
      .find('.sidebar__sources .sidebar__enabled')
      .find('.sidebar__element__button-hidden').trigger('click');
    expect(feed.actions.setEnablePublication)
      .toBeCalledWith(expect.anything(), { index: 0, enabled: false }, undefined);
    done();
  });
});

it('should commit "setEnablePublication" when adding publication', (done) => {
  const wrapper = mount(DaSidebar, { store, localVue });
  wrapper.vm.filterChecked = false;
  wrapper.vm.$nextTick(() => {
    wrapper
      .find('.sidebar__sources .sidebar__disabled')
      .find('.sidebar__element__button-hidden').trigger('click');
    expect(feed.actions.setEnablePublication)
      .toBeCalledWith(expect.anything(), { index: 1, enabled: true }, undefined);
    done();
  });
});

it('should set enabledTags according to state', () => {
  const wrapper = shallowMount(DaSidebar, { store, localVue });
  expect(wrapper.vm.enabledTags).toEqual([{
    'name': 'javascript',
    'enabled': true,
  }, {
    'name': 'linux',
    'enabled': true,
  }]);
});

it('should set disabledTags according to state', () => {
  const wrapper = shallowMount(DaSidebar, { store, localVue });
  expect(wrapper.vm.disabledTags).toEqual([{
    'name': 'startup',
    'enabled': false,
  }, {
    'name': 'product',
    'enabled': false,
  }]);
});

it('should dispatch "setEnableTag" when removing tag', (done) => {
  const wrapper = mount(DaSidebar, { store, localVue });
  wrapper.vm.filterChecked = true;
  wrapper.vm.$nextTick(() => {
    wrapper
      .find('.sidebar__tags .sidebar__enabled')
      .find('.sidebar__element__button-hidden').trigger('click');
    expect(feed.actions.setEnableTag)
      .toBeCalledWith(expect.anything(), { tag: feed.state.tags[0], enabled: false }, undefined);
    done();
  });
});

it('should commit "setEnableTag" when adding tag', (done) => {
  const wrapper = mount(DaSidebar, { store, localVue });
  wrapper.vm.filterChecked = true;
  wrapper.vm.$nextTick(() => {
    wrapper
      .find('.sidebar__tags .sidebar__disabled')
      .find('.sidebar__element__button-hidden').trigger('click');
    expect(feed.actions.setEnableTag)
      .toBeCalledWith(expect.anything(), { tag: feed.state.tags[2], enabled: true }, undefined);
    done();
  });
});

it('should activate request source form', (done) => {
  user.state.profile = { name: 'John' };
  const wrapper = mount(DaSidebar, { store, localVue });
  wrapper.vm.$nextTick(() => {
    expect(wrapper.vm.requestActive).toEqual(false);
    wrapper
      .find('.sidebar__sources__act-req').trigger('click');
    expect(wrapper.vm.requestActive).toEqual(true);
    done();
  });
});

describe('SUBMIT REQUEST', () => {
  let wrapper;
  beforeEach(() => {
    user.state.profile = { name: 'John' };
    wrapper = mount(DaSidebar, { store, localVue });
    wrapper.vm.requestActive = true;
  });

  it('should enable submit when input is a valid url', (done) => {
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.disableSubmit).toEqual(true);
      wrapper.vm.$refs.request.value = REQUEST_URL;
      wrapper.find('.sidebar__sources .sidebar__input').trigger('input');
      expect(wrapper.vm.disableSubmit).toEqual(false);
      done();
    });
  });

  it('should disable request input while the request is processed', (done) => {
    // Simulate delay
    contentService.requestPublication.mockReturnValue(new Promise(
      (resolve) => setTimeout(() => resolve(), 3000))
    );
    wrapper.vm.$nextTick(() => {
      wrapper.vm.$refs.request.value = REQUEST_URL;
      const sideBarSources = wrapper.find('.sidebar__sources');
      const sideBarSourcesInput = sideBarSources.find('.sidebar__input');
      sideBarSourcesInput.trigger('input');
      sideBarSources.find('.sidebar__sources__submit').trigger('click');
      expect(sideBarSourcesInput.attributes('disabled')).toBeTruthy();
      done();
    });
  });
});

it('should dispatch "setFilter" with publication filter', (done) => {
  const wrapper = mount(DaSidebar, { store, localVue });
  wrapper.vm.$nextTick(() => {
    wrapper
      .find('.sidebar__sources .sidebar__disabled')
      .find('.sidebar__element__button').trigger('click');
    expect(feed.actions.setFilter)
      .toBeCalledWith(expect.anything(), {
        type: 'publication',
        info: feed.state.publications[1],
      }, undefined);
    done();
  });
});

it('should dispatch "setFilter" with tag filter', (done) => {
  const wrapper = mount(DaSidebar, { store, localVue });
  wrapper.vm.filterChecked = true;
  wrapper.vm.$nextTick(() => {
    wrapper
      .find('.sidebar__tags .sidebar__disabled')
      .find('.sidebar__element__button').trigger('click');
    expect(feed.actions.setFilter)
      .toBeCalledWith(expect.anything(), {
        type: 'tag',
        info: feed.state.tags[2],
      }, undefined);
    done();
  });
});

it('should search for tags and update the tag list', (done) => {
  const wrapper = mount(DaSidebar, { store, localVue });
  wrapper.vm.filterChecked = true;
  wrapper.vm.$nextTick(() => {
    contentService.searchTags.mockReturnValue({
      query: 'java',
      hits: [{ name: 'javascript' }, { name: 'java' }],
    });
    wrapper.vm.$refs.search.value = 'java';
    wrapper.find('.sidebar__search .sidebar__input').trigger('input');
    wrapper.vm.$nextTick(() => {
      expect(contentService.searchTags).toBeCalledTimes(1);
      expect(contentService.searchTags).toBeCalledWith('java');
      expect(wrapper.vm.activeTags).toEqual([
        { name: 'javascript', enabled: true },
        { name: 'java', enabled: false },
      ]);
      done();
    });
  });
});

it('should get the right placeholder name for the search input', async () => {
  let expected;

  const wrapper = mount(DaSidebar, { store, localVue });
  expected = 'Search Sources';

  expect(wrapper.vm.searchPlaceholder).toBe(expected);

  expected = 'Search Tags';
  wrapper.vm.filterChecked = true;
  await wrapper.vm.$nextTick();

  expect(wrapper.vm.searchPlaceholder).toBe(expected);
});

describe('SEARCH: Publications', () => {
  it('should filter publications depending on input value', async () => {
    const wrapper = mount(DaSidebar, { store, localVue });
    const disabledPublicationsSelector = '.sidebar__sources .sidebar__disabled .sidebar__element.btn';
    const enabledPublicationsSelector = '.sidebar__sources .sidebar__enabled .sidebar__element.btn';
    const expected = feed.state.publications[1];
    wrapper.vm.query = 'all';

    await wrapper.vm.$nextTick();

    expect(wrapper.vm.filteredPublications[0].name).toBe(expected.name);
    expect(wrapper.findAll('.sidebar__sources .sidebar__disabled .sidebar__element.btn').length).toBe(1);

    wrapper.vm.query = 'a';

    await wrapper.vm.$nextTick();

    expect(wrapper.vm.disabledPubs.length).toBe(2);
    expect(wrapper.findAll(disabledPublicationsSelector).length).toBe(2);

    expect(wrapper.vm.enabledPubs.length).toBe(2);
    expect(wrapper.findAll(enabledPublicationsSelector).filter(el => !el.classes('sidebar__sources__act-req')).length).toBe(2);
  });

  it('should display all the publications if the input value is empty', async () => {
    const wrapper = mount(DaSidebar, { store, localVue });
    const expected = feed.state.publications;
    wrapper.vm.query = '';

    await wrapper.vm.$nextTick();

    expect(wrapper.vm.filteredPublications).toBe(expected);
  });

  it('should display no publications if the input value does not match at all', async () => {
    const wrapper = mount(DaSidebar, { store, localVue });
    wrapper.vm.query = 'abcdefgh';

    await wrapper.vm.$nextTick();

    expect(wrapper.vm.filteredPublications.length).toBe(0);
  });
});
