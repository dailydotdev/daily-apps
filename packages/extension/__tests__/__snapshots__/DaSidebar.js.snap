// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEARCH: Publications should display all the publications if the input value is empty 1`] = `
Array [
  Object {
    "enabled": true,
    "id": "airbnb",
    "image": "https://res.cloudinary.com/daily-now/image/upload/t_logo,f_auto/v1/logos/airbnb",
    "name": "Airbnb Engineering",
  },
  Object {
    "enabled": false,
    "id": "alligator",
    "image": "https://res.cloudinary.com/daily-now/image/upload/t_logo,f_auto/v1/logos/alligator",
    "name": "Alligator",
  },
  Object {
    "enabled": false,
    "id": "angular",
    "image": "https://res.cloudinary.com/daily-now/image/upload/t_logo,f_auto/v1/logos/angular",
    "name": "Angular",
  },
  Object {
    "enabled": true,
    "id": "aws",
    "image": "https://res.cloudinary.com/daily-now/image/upload/t_logo,f_auto/v1/logos/aws",
    "name": "AWS",
  },
]
`;
